FVT: The "Liberty: Start" commands are failing in the initial run when the parent directory for Liberty project(s) is open in the Project window

When we execute dev/devc command for the first time it is failing because of 'No such file/directory' error.
Os- Redhat
LTI version: 24.0.6


Logs :
024.1.2/idea-IC-241.17011.79/plugins/maven/lib/maven3/bin/mvn" io.openliberty.tools:liberty-maven-plugin:devc 
cd "/root/Documents/test-one/app-name-testone"
[root@c2826v1 test-one]# "/root/Documents/24.0.6-test/ideaIC-2024.1.2/idea-IC-241.17011.79/plugins/maven/lib/maven3/bin/mvn" io.openliberty.tools:liberty-maven-plugin:devc 
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< org.apache.maven:standalone-pom >-------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.269 s
[INFO] Finished at: 2024-06-05T00:57:20-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.openliberty.tools:liberty-maven-plugin:3.9:devc (default-cli): Goal requires a project to execute but there is no POM in this directory (/root/Documents/test-one). Please verify you invoked Maven from the correct directory. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MissingProjectException

---

- [ ] Resolve Mac scenario
- [ ] Resolve Windows scenario: https://github.com/OpenLiberty/liberty-tools-intellij/issues/797#issuecomment-2150903563
- [ ] Resolve multi module scenario: https://github.com/OpenLiberty/liberty-tools-intellij/issues/807

--- a/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevRunTestsAction.java
+++ b/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevRunTestsAction.java
@@ -31,11 +31,11 @@ public class LibertyDevRunTestsAction extends LibertyGeneralAction {
     protected void executeLibertyAction(LibertyModule libertyModule) {
         Project project = libertyModule.getProject();
         VirtualFile buildFile = libertyModule.getBuildFile();
-        String runTestsCommand = " ";
         ShellTerminalWidget widget = getTerminalWidgetWithFocus(false, project, buildFile, getActionCommandName());
         if (widget == null) {
             return;
         }
+        String runTestsCommand = " ";
         LibertyActionUtil.executeCommand(widget, runTestsCommand);
     }
 

--- a/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevStartAction.java
+++ b/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevStartAction.java
@@ -82,8 +82,7 @@ public class LibertyDevStartAction extends LibertyGeneralAction {
         }
 
         String cdToProjectCmd = "cd \"" + buildFile.getParent().getPath() + "\"";
-        LibertyActionUtil.executeCommand(widget, cdToProjectCmd);
-        LibertyActionUtil.executeCommand(widget, startCmd);
+        LibertyActionUtil.executeCommand(widget, cdToProjectCmd, startCmd);
         if (libertyModule.isDebugMode() && debugPort != -1) {
             // Create remote configuration to attach debugger
             debugHandler.createAndRunDebugConfiguration(libertyModule, debugPort);

--- a/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevStartContainerAction.java
+++ b/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevStartContainerAction.java
@@ -51,7 +51,6 @@ public class LibertyDevStartContainerAction extends LibertyGeneralAction {
             return;
         }
         String cdToProjectCmd = "cd \"" + buildFile.getParent().getPath() + "\"";
-        LibertyActionUtil.executeCommand(widget, cdToProjectCmd);
-        LibertyActionUtil.executeCommand(widget, startCmd);
+        LibertyActionUtil.executeCommand(widget, cdToProjectCmd, startCmd);
     }
 }

--- a/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevStopAction.java
+++ b/src/main/java/io/openliberty/tools/intellij/actions/LibertyDevStopAction.java
@@ -32,10 +32,10 @@ public class LibertyDevStopAction extends LibertyGeneralAction {
         Project project = libertyModule.getProject();
         VirtualFile buildFile = libertyModule.getBuildFile();
         ShellTerminalWidget widget = getTerminalWidgetWithFocus(false, project, buildFile, getActionCommandName());
-        String stopCmd = "q";
         if (widget == null) {
             return;
         }
+        String stopCmd = "q";
         LibertyActionUtil.executeCommand(widget, stopCmd);
     }
 }

--- a/src/main/java/io/openliberty/tools/intellij/util/LibertyActionUtil.java
+++ b/src/main/java/io/openliberty/tools/intellij/util/LibertyActionUtil.java
@@ -39,6 +39,10 @@ public class LibertyActionUtil {
             int i = 0;
             try {
                 while (widget.getProcessTtyConnector() == null) {
+                    if (i > 100) {
+                        LOGGER.warn("Time out waiting to execute command: " + cmd1);
+                        return;
+                    }
                     LOGGER.warn("Waiting for cd to execute: " + i++);
                     Thread.sleep(100);
                 }

--- a/src/main/java/io/openliberty/tools/intellij/util/LibertyActionUtil.java
+++ b/src/main/java/io/openliberty/tools/intellij/util/LibertyActionUtil.java
@@ -9,6 +9,7 @@
  *******************************************************************************/
 package io.openliberty.tools.intellij.util;
 
+import com.intellij.openapi.application.ApplicationManager;
 import com.intellij.openapi.diagnostic.Logger;
 import com.jediterm.terminal.TtyConnector;
 import org.jetbrains.plugins.terminal.ShellTerminalWidget;
@@ -19,6 +20,35 @@ import java.nio.charset.StandardCharsets;
 
 public class LibertyActionUtil {
 
+    static Logger LOGGER = Logger.getInstance(LibertyActionUtil.class);
+    /**
+     * Send the given two commands to the given ShellTerminalWidget
+     * but ensure the first command has at least started executing
+     * before typing the second command on the terminal.
+     *
+     * @param widget
+     * @param cmd1
+     * @param cmd2
+     */
+    public static void executeCommand(ShellTerminalWidget widget, String cmd1, String cmd2) {
+        // Perform these commands on the same pooled thread or else the event thread will be blocked
+        ApplicationManager.getApplication().executeOnPooledThread(() -> {
+            executeCommand(widget, cmd1);
+            // Do not run the second command until the execution of the first command has begun.
+            // This is required because IntelliJ batches commands and runs them out of order.
+            int i = 0;
+            try {
+                while (widget.getProcessTtyConnector() == null) {
+                    LOGGER.warn("Waiting for cd to execute: " + i++);
+                    Thread.sleep(100);
+                }
+            } catch (InterruptedException x) {
+                LOGGER.error(String.format("Interrupted waiting to execute command: %s", cmd1), x);
+            }
+            executeCommand(widget, cmd2);
+        });
+    }
+
     /**
      * Send the given command to the given ShellTerminalWidget
      *
@@ -26,12 +56,11 @@ public class LibertyActionUtil {
      * @param cmd
      */
     public static void executeCommand(ShellTerminalWidget widget, String cmd) {
-        Logger LOGGER = Logger.getInstance(LibertyActionUtil.class);
         try {
             widget.grabFocus();
             TtyConnector connector = widget.getTtyConnector();
             if (connector == null) {
-                // new terminal, use build in execute command function
+                // new terminal, use built in execute command function
                 widget.executeCommand(cmd);
                 return;
             }

